File: scripts/irr-verify/verify.R
# Recompute every statistic in R with the irr package (kripp.alpha,
# kappam.fleiss, kappa2) plus a from-scratch mean pairwise agreement.
# Run: Rscript scripts/irr-verify/verify.R <outdir>   (IRR_RLIB adds a library path)
args <- commandArgs(trailingOnly = TRUE)
out <- if (length(args)) args[1] else "scripts/irr-verify/out"
rlib <- Sys.getenv("IRR_RLIB")
if (nzchar(rlib)) .libPaths(c(rlib, .libPaths()))
suppressPackageStartupMessages({ library(irr); library(irrCAC); library(jsonlite) })

# irr::kripp.alpha weights each unit's pairs by 1/(m - 1) only when the matrix
# holds an NA; with no NA at all it uses a weight of 1 (`mc <- rep(1, ...)`),
# which is right for two raters and wrong for three or more. Appending a rater
# who coded nothing forces the general branch without adding a single pair, so
# the function's own arithmetic then gives Krippendorff's value. The unpatched
# result is kept as *Raw so the comparison can document the discrepancy.
force_general <- function(m) if (any(is.na(m))) m else rbind(m, NA_real_)

ds <- fromJSON(file.path(out, "datasets.json"), simplifyVector = FALSE)

mean_pairwise <- function(m) {
  agree <- 0; pairs <- 0
  for (i in seq_len(ncol(m))) {
    v <- m[, i]; v <- v[!is.na(v)]
    if (length(v) < 2) next
    cb <- combn(v, 2)
    pairs <- pairs + ncol(cb)
    agree <- agree + sum(cb[1, ] == cb[2, ])
  }
  if (pairs == 0) NA else agree / pairs
}
num <- function(x) if (is.null(x) || length(x) == 0 || is.na(x) || !is.finite(x)) NA else as.numeric(x)

res <- lapply(ds, function(d) {
  m <- matrix(NA_real_, nrow = d$R, ncol = d$N)
  for (r in seq_len(d$R)) for (i in seq_len(d$N)) {
    v <- d$cells[[r]][[i]]
    if (!is.null(v)) m[r, i] <- v + 1
  }
  ka <- function(method) num(tryCatch(kripp.alpha(force_general(m), method)$value, error = function(e) NA))
  kaRaw <- function(method) num(tryCatch(kripp.alpha(m, method)$value, error = function(e) NA))
  # Gwet's irrCAC: subjects x raters, NA for missing. "unweighted" is nominal alpha;
  # "quadratic" weights are compared against the interval metric as a soft check.
  cac <- function(w) num(tryCatch(krippen.alpha.raw(t(m), weights = w)$est$coeff.val, error = function(e) NA))
  counts <- colSums(!is.na(m)); use <- counts >= 2
  fl <- NA; flCac <- NA
  if (any(use) && length(unique(counts[use])) == 1) {
    mm <- unique(counts[use])
    subj <- do.call(rbind, lapply(which(use), function(i) { v <- m[, i]; v[!is.na(v)] }))
    fl <- num(tryCatch(kappam.fleiss(subj)$value, error = function(e) NA))
    flCac <- num(tryCatch(fleiss.kappa.raw(subj)$est$coeff.val, error = function(e) NA))
  } else flCac <- NA
  pairs <- list()
  for (a in seq_len(d$R - 1)) for (b in (a + 1):d$R) {
    x <- cbind(m[a, ], m[b, ]); x <- x[complete.cases(x), , drop = FALSE]
    k <- if (nrow(x) >= 2) num(tryCatch(kappa2(x)$value, error = function(e) NA)) else NA
    po <- if (nrow(x) >= 2) mean(x[, 1] == x[, 2]) else NA
    pairs[[length(pairs) + 1]] <- list(a = a - 1, b = b - 1, kappa = k, po = po, n = nrow(x))
  }
  perCode <- lapply(seq_len(d$K), function(k) {
    b <- ifelse(is.na(m), NA, ifelse(m == k, 1, 2))
    num(tryCatch(kripp.alpha(force_general(b), "nominal")$value, error = function(e) NA))
  })
  list(id = d$id, complete = !any(is.na(m)),
       alphaNominal = ka("nominal"), alphaOrdinal = ka("ordinal"), alphaInterval = ka("interval"),
       alphaNominalRaw = kaRaw("nominal"), alphaOrdinalRaw = kaRaw("ordinal"), alphaIntervalRaw = kaRaw("interval"),
       cacNominal = cac("unweighted"), cacQuadratic = cac("quadratic"), cacFleiss = flCac,
       fleiss = fl, meanPairwise = num(mean_pairwise(m)), pairs = pairs, perCode = perCode)
})
write_json(res, file.path(out, "r.json"), auto_unbox = TRUE, digits = NA, na = "null")
cat("wrote R results for", length(res), "datasets to", file.path(out, "r.json"), "\n")

File: scripts/irr-verify/verify.py
"""Recompute the statistics in Python: Krippendorff's alpha (nominal, ordinal,
interval) with the `krippendorff` package, and, when IRR_ANALYSIS_PY points at
the classroom script irr_analysis.py, its alpha, Fleiss, Cohen, mean pairwise
agreement, one-against-all alpha, and a bootstrap with numpy's generator.

Run: python3 scripts/irr-verify/verify.py <outdir>
"""
import importlib.util
import json
import math
import os
import sys

import numpy as np
import krippendorff

out = sys.argv[1] if len(sys.argv) > 1 else "scripts/irr-verify/out"
ds = json.load(open(os.path.join(out, "datasets.json")))

oz = None
ap = os.environ.get("IRR_ANALYSIS_PY")
if ap and os.path.exists(ap):
    spec = importlib.util.spec_from_file_location("irr_analysis", ap)
    oz = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(oz)


def nz(v):
    if v is None:
        return None
    v = float(v)
    return None if (math.isnan(v) or math.isinf(v)) else v


res = []
for d in ds:
    K, R, N = d["K"], d["R"], d["N"]
    arr = np.full((R, N), np.nan)
    for r in range(R):
        for i in range(N):
            v = d["cells"][r][i]
            if v is not None:
                arr[r, i] = v + 1

    def ka(level):
        try:
            return nz(krippendorff.alpha(reliability_data=arr, level_of_measurement=level))
        except Exception:
            return None

    rec = {"id": d["id"], "alphaNominal": ka("nominal"), "alphaOrdinal": ka("ordinal"), "alphaInterval": ka("interval")}

    if oz is not None:
        codes = tuple(str(k + 1) for k in range(K))
        oz.CODES = codes
        cols = [[str(int(arr[r, i])) for r in range(R) if not np.isnan(arr[r, i])] for i in range(N)]
        rows = [[(str(int(arr[r, i])) if not np.isnan(arr[r, i]) else None) for i in range(N)] for r in range(R)]
        rec["ozAlphaNominal"] = nz(oz.krippendorff_alpha(cols))
        fk, note = oz.fleiss_kappa(cols)
        rec["ozFleiss"] = nz(fk)
        rec["ozFleissIncomplete"] = bool(note and "varies" in note)
        rec["ozMeanPairwise"] = nz(oz.mean_pairwise(cols))
        pairs = []
        for a in range(R):
            for b in range(a + 1, R):
                ck = oz.cohen_kappa(rows[a], rows[b])
                pairs.append({"a": a, "b": b, "kappa": nz(ck["kappa"]) if ck else None,
                              "po": nz(ck["po"]) if ck else None, "n": ck["n"] if ck else 0})
        rec["ozPairs"] = pairs
        rec["ozPerCode"] = [nz(oz.alpha_binary(cols, codes[k])) for k in range(K)]
        if d["id"] < 9:
            usable = [c for c in cols if len(c) >= 2]
            rng = np.random.default_rng(7)
            vals = []
            for _ in range(6000):
                idx = rng.integers(0, len(usable), len(usable))
                a = oz.krippendorff_alpha([usable[j] for j in idx])
                if a is not None:
                    vals.append(a)
            vals.sort()
            rec["ozBoot"] = {"lo": vals[int(0.025 * len(vals))], "hi": vals[int(0.975 * len(vals))]}
    res.append(rec)

json.dump(res, open(os.path.join(out, "py.json"), "w"))
print(f"wrote Python results for {len(res)} datasets to {out}/py.json" + ("" if oz else " (classroom script not loaded)"))
